### PR TITLE
Cascaded spill merge and re-spill

### LINKED ISSUE #14692: A complete solution for stable and safe sort with spill
### Is your feature request related to a problem or challenge?

This is a follow-up for https://github.com/apache/datafusion/pull/14644:

See details in https://github.com/apache/datafusion/pull/14644#discussion_r1957155614 cc @alamb @Kontinuation 

We need a complete solution about stable sort with spill, several issues until now:

1. Ensuring we can always spill data (now spilling will sometimes fail if we run out of memory to sort the batches in)
2. Ensuring that we can always merge the data that was spilled, even if it had a really wide fanout (like 1000 of spill files)

The think problem 1 could be solved by potentially spilling unsorted batches (and then sorting them separately). This would be less efficient (read/write some tuples twice but would work.

Problem 2 could use the multi-pass merge:

We can implement a more complicated multi-stage merging phase which merges a small portion of streams each time, and perform multiple rounds of merges to produce the final merged stream. 

### Describe the solution you'd like

_No response_

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by 2010YOUY01**
## Which issue does this PR close?

- Closes #14692

### Background for memory-limited sort execution
See figures in https://github.com/apache/datafusion/blob/main/datafusion/physical-plan/src/sorts/sort.rs

### Current limitation for reading back spill files
Let's say the memory buffer can only hold 10 batches, but there are 900 files spilled, the current implementation will try to merge 900 files at once, and fail the query.
However, this scenario is possible to continue if it can only merge 10 files at a time, re-spill, until there are only less than 10 spill files in total, and finally read them back and merge them as the output.

### High-level approach of this PR
Added one configuration option for max spill merge degree (haven't implemented now, in POC it's a hard-coded `const MAX_SPILL_MERGE_DEGREE` for simplicity)
At the final stage of external sort, there are initial spill files to merge, perform multi-pass read-merge-respill operation, the number of merged spill file in the next pass is decided by the closet power of `MAX_SPILL_MERGE_DEGREE`

Example:
Initial spill files to merge: 900
max merge degree: 10

pass1: merge 900 files to 100 (closet power of 10) files
pass2: 100 -> 10
pass3: 10 -> final single output

Inside each pass the number of files to merge in each step is split as even as possible, while is always <= max merge degree, see details in the implementation.

## What changes are included in this PR?

Updated the `sort.rs` for multi-pass reading spill. The entry point for the described logic is function `merge_spilled_files_multi_pass()`

## Are these changes tested?

To be done, I think it's adding varying `max_spill_merge_degree` to the sort fuzzer.

## Are there any user-facing changes?

No

### CODE REVIEW

**Comment by rluvaton** [comment]
BTW, row_hash uses the sort preserving merge stream as well and has similar problem, I think this should be a solution outside the sort exec

**Comment by rluvaton** [comment]
Also, to have a fully working sort, you need to spill in
 https://github.com/apache/datafusion/blob/362fcdfc7b9e00cb6126a0cbc41c9abb2637c563/datafusion/physical-plan/src/sorts/builder.rs#L74

In case the memory reservation is failing

**Comment by 2010YOUY01** [comment]
This PR and https://github.com/apache/datafusion/pull/15608 both implemented multi-level merge for `SortExec` for different purposes:
### This PR
- This PR wants to let memory-limit sort queries be able to run even if the memory budget is very tight (i.e. num-spill-files * batch-size > memory limit)
- Always re-spill for each merge step
### #15608
- Reduces merge degree for performance (reading spills will stall for a shorter amount of time)
- Never re-spill

I think we should refine the existing PR to be:
1. Prioritize stable execution of memory-limited queries over performance.
    - I think the optimizations mentioned below are somewhat complex. We should first resolve the remaining known correctness issues in external sort, strengthen the tests, and then proceed with later optimizations more confidently.
2. Extensible for future performance optimization
    - When the memory budget allows, don't always re-spill
    - Consider pre-fetching future spill reads to avoid blocking read
    - For other steps that require merging in `SortExec`, the multi-pass merging utility should be reusable for performance: For example if we have enough memory to buffer all input buffers, it should be able to do multi-level merging.
      (The first two points is related to https://github.com/apache/datafusion/issues/15323, the third point is tracked in https://github.com/apache/datafusion/issues/7181)


To summarize, I think this PR needs to be restructured to make future optimizations easier to implement. I don’t have a solid idea yet, so I’ll keep thinking and also wait to hear more opinions.

File: datafusion/physical-plan/src/sorts/sort.rs [hunk 1]
```diff
@@ -535,56 +457,262 @@ impl ExternalSorter {
...
+        debug!("Combining spilled files");
+
+        // ==== Write to a single merged spill file ====
+        let merged_spill_file = self.write_stream_to_spill_file(spm_stream).await?;
+
+        Ok(merged_spill_file)
+    }
+
+    /// Maximum number of spill files to merge in a single pass
+    const MAX_SPILL_MERGE_DEGREE: usize = 8;
```
[the remark this answers is not in the document]
Comment (2010YOUY01) [reply datafusion/physical-plan/src/sorts/sort.rs:579]: Yeah, it's only for POC.

**Comment by 2010YOUY01** [comment]
> Also, to have a fully working larger than memory sort, you need to spill in
> 
> https://github.com/apache/datafusion/blob/362fcdfc7b9e00cb6126a0cbc41c9abb2637c563/datafusion/physical-plan/src/sorts/builder.rs#L74
> 
> In case the memory reservation is failing

Could you elaborate? I don't get it.

File: datafusion/physical-plan/src/sorts/sort.rs [hunk 2]
```diff
@@ -535,56 +457,262 @@ impl ExternalSorter {
...
+            .build()?;
+
+        Ok(Box::pin(RecordBatchStreamAdapter::new(
+            Arc::clone(&self.schema),
+            spm_stream,
+        )))
+    }
+
+    /// Recursively merges and re-spills files until the number of spill files is ≤ MAX_SPILL_MERGE_DEGREE
+    async fn recursively_merge_spill_files(
```
Comment (qstommyshu) [inline datafusion/physical-plan/src/sorts/sort.rs:660]: I find this name misleading, looks like this is not a recursive function.

File: datafusion/physical-plan/src/sorts/sort.rs [hunk 3]
```diff
@@ -535,56 +457,262 @@ impl ExternalSorter {
...
+                }
+
+                self.spill_manager.read_spill_as_stream(spill_file)
+            })
+            .collect::<Result<Vec<_>>>()?;
+
+        let sort_exprs: LexOrdering = self.expr.iter().cloned().collect();
+
+        // ==== Doing sort-preserving merge on input partially sorted streams ====
+        let spm_stream = StreamingMergeBuilder::new()
```
Comment (qstommyshu) [inline datafusion/physical-plan/src/sorts/sort.rs:560]: I wonder if this StreamingMergeBuilder uses heap under the hood, using heap is a common method to optimize external merge sort performance

**Review by qstommyshu — CHANGES_REQUESTED** [review]

File: datafusion/physical-plan/src/sorts/sort.rs (hunk 2 above)
Comment (2010YOUY01) [reply datafusion/physical-plan/src/sorts/sort.rs]: Good point, updated. I changed my mind midway through implementing this function.

File: datafusion/physical-plan/src/sorts/sort.rs (hunk 3 above)
Comment (2010YOUY01) [reply datafusion/physical-plan/src/sorts/sort.rs:560]: Yes, it's an in-house implementation of loser-tree heap. If we don't limit the degree of merge, for large sort queries, this step is the bottleneck now, maybe there is some room to optimize inside 🤔

Comment (alamb) [reply datafusion/physical-plan/src/sorts/sort.rs:560]: It is already pretty well optimized (not that it couldn't be made better) but there isn't a lot of low hanging fruit in my opinion

File: datafusion/physical-plan/src/sorts/sort.rs [hunk 4]
```diff
@@ -535,56 +457,262 @@ impl ExternalSorter {
...
+        // initial spill files count: 30
+        // max merge degree: 4
+        // pass 1: merge 30 files into 16(4^2) files
+        // pass 2: merge 16 files into 4(4^1) files
+        // pass 3: now the number of spill files is <= max merge degree: merge them into a single sorted stream
+        while spill_files_cur_pass.len() > Self::MAX_SPILL_MERGE_DEGREE {
+            let log_base = Self::MAX_SPILL_MERGE_DEGREE as f64;
+            let num_files = spill_files_cur_pass.len() as f64;
+            let num_passes = num_files.log(log_base).ceil() as usize;
+            let next_pass_merge_degree = log_base.powi((num_passes - 1) as i32);
```
Comment (qstommyshu) [inline datafusion/physical-plan/src/sorts/sort.rs:679]: Maybe we can add a comment to explain what is `next_pass_merge_degree`?

File: datafusion/physical-plan/src/sorts/sort.rs [hunk 5]
```diff
@@ -535,56 +457,262 @@ impl ExternalSorter {
...
+    ///     merged_stream: batch(100), batch(100), batch(100), batch(100)
+    /// ```
+    async fn merge_spilled_files_multi_pass(
+        &mut self,
+    ) -> Result<SendableRecordBatchStream> {
+        // ──────────────────────────────────────────────────────────────────────
+        // Edge cases
+        // ──────────────────────────────────────────────────────────────────────
+        if self.finished_spill_files.is_empty() {
+            return internal_err!("No spilled files to merge");
```
Comment (qstommyshu) [inline datafusion/physical-plan/src/sorts/sort.rs:610]: I think the error here should be different from the case `part_tially_sorted_stream.is_empty()`, that way it is easier to identify error if there is any bugs.

File: datafusion/physical-plan/src/sorts/sort.rs [hunk 6]
```diff
@@ -535,56 +457,262 @@ impl ExternalSorter {
...
+    ///     spill_file_4: batch(100)
+    ///
+    /// After pass 1:
+    ///     spill_file_1: batch(100), batch(100)
+    ///     spill_file_2: batch(100), batch(100)
+    ///
+    /// After pass 2:
+    ///     merged_stream: batch(100), batch(100), batch(100), batch(100)
+    /// ```
+    async fn merge_spilled_files_multi_pass(
```
Comment (qstommyshu) [inline datafusion/physical-plan/src/sorts/sort.rs:603]: minor issue, the naming for `multi_pass` is not consistent, it is `multipass` below.

**Review by qstommyshu — COMMENTED** [review]
Just some minor issues, the algorithm itself looks good to me in general. I can take a closer look in details if needed (I'm not familiar with this part of the codebase yet, but I'll try my best to provide good review comments).

And some other thoughts:
1. This is a pretty complicated program, maybe we should write some unit tests to make sure it doesn't break for future modifications?
2. One idea to improve the performance is to dynamic calculate the optimal merge degree based on file size and memory size, or maybe multi-thread the merge phase (not sure if it is feasible)

File: datafusion/physical-plan/src/sorts/sort.rs [hunk 7]
```diff
@@ -535,56 +457,262 @@ impl ExternalSorter {
         // reserved again for the next spill.
         self.merge_reservation.free();
 
-        let mut sorted_stream =
+        let sorted_stream =
             self.in_mem_sort_stream(self.metrics.baseline.intermediate())?;
+        debug!("SPM stream is constructed");
```
Comment (qstommyshu) [inline datafusion/physical-plan/src/sorts/sort.rs:462]: Clean up debug logs if they are not needed?

**Review by qstommyshu — APPROVED** [review]

**Review by alamb — COMMENTED** [review]
This is very cool -- thank you @2010YOUY01  and @rluvaton and @qstommyshu 

I think in general (can do as a follow on PR) we will need to introduce some parallelism as well to really maximize performance. 

Specifically, the merge operation is fundamentally single threaded (the hot loop in the merge). Thus I would expect merging one set of 10 files to likely be CPU bottlenecked

So we probably would need to try and merge multiple sets of 10 files in parallel (to multiple different output files) before we either bottlenecked on CPU or on I/O throughput

What I think would really help make progress in this area is A benchmark. I filed a ticket to track this issue:
- https://github.com/apache/datafusion/issues/15664

**Comment by 2010YOUY01** [comment]
Thank you all for the review!

@qstommyshu I agree with the implementation-level feedbacks. I will address them in the refactor.

@alamb Regarding parallel merging: I was thinking if `max_spill_perge_degree` configured to 10, than the memory is limited so that in each partition, we can only hold 10 batches at the same time, so parallel merging is not possible in this case.
However, @rluvaton 's PR has inspired me that, it's possible each operator is able to hold 100 batches under the memory limit at the same time, but we might still want to merge them 10 at a time for performance.

I think the next steps are
1. Contribute benchmarks for external sort.
2. Refactor this PR to avoid always re-spill, also do parallel merging when possible.

**Comment by 2010YOUY01** [comment]
> And some other thoughts:
> 
> 1. This is a pretty complicated program, maybe we should write some unit tests to make sure it doesn't break for future modifications?

I'll try to do most of the testing and cover edge cases in integration tests at https://github.com/apache/datafusion/blob/main/datafusion/core/tests/fuzz_cases/sort_fuzz.rs and https://github.com/apache/datafusion/blob/main/datafusion/core/tests/fuzz_cases/sort_query_fuzz.rs, instead of doing extensive UTs.

I think we should promote tests to a higher level (SQL) when possible, because that API is much more stable and easier to manage. If a feature is tested mostly through unit tests, and someone later refactors the component away, those tests are likely to get lost—they might assume the feature is already covered by integration tests.

I first heard this idea in a talk by the DuckDB developers https://youtu.be/BgC79Zt2fPs?si=WiziGqJ8Dlz6-MMW

**Comment by 2010YOUY01** [comment]
I didn't implement the parallel merge optimization for now, my major concern is: this optimization requires one extra configuration, and users have to learn and correctly set 2 configs for each individual query, to enable the most efficient cascaded spill merge execution (see the below `intended optimization` section for what's those 2 configs), which is not ideal.
So I'd like to defer the implementation for a bit, to think about if there are any simpler approaches (or maybe by collecting stats internally and auto-tune those related configs)
Also, I think the current implementation is good enough to cover common cases (I did a rough estimation, sorting TPCH-SF1000 lineitem table with 16GB of memory only requires one round of re-spill)

Here is the optimization that I originally thought about, I'll put them into a separate issue if it makes sense.
### Example scenario
For one partition's `SortExec`, 100 runs are spilled, and we set `spill_max_spill_merge_degree` to 4
### Current Implementation
image
Each time it merges 4 existing spills into one combined spill file, until there are <= 4 spills total, the final result can be produced.
For each entry, the number of re-spill will be $floor(\log_4 100)$ = 3

### Intended optimization
If the memory pool is enough to hold more buffers at a time (while `spill_max_spill_merge_degree` is still limited to 4, in case the merge-degree is too large and hurt performance in some cases)
One additional config will be introduced `sort_buffer_batch_capacity`, and set to `16` in the above example, the execution will look like:
image
Then, inside each merge step, 16 spill files will be combined and re-spill. Each entry only need to be re-spilled for $floor(\log_{16} 100)$ = 1 time.
With this approach, we can achieve an optimal re-spill count, and also enable parallel merge.

**Comment by 2010YOUY01** [comment]
Benchmark results:
(I think there is no significant regression for an extra round of re-spill, if it's running on a machine with fast SSDs)

### Environment
MacBook Pro with m4-pro chip (disk bandwidth is around 8000MB/s)
### Sorting 'thin' table
1. Run datafusion-cli with `cargo run --profile release-nonlto -- --mem-pool-type fair -m 100M`
2. Execute `explain analyze select * from generate_series(1, 1000000000) as t1(v1) order by v1;`

Main: 37s (merge ~170 spill files at once)
PR (with `sort_max_spill_merge_degree = 16`, and there is one round of re-spill): 43s
PR (with `sort_max_spill_merge_degree = 10`, two rounds of re-spill): 49s
### Sorting 'fat' table
Run `sort_tpch` benchmark q7
```
        // Q7: 3 sort keys {(INTEGER, 7), (BIGINT, 10k), (BIGINT, 1.5M)} + 12 all other columns
        r#"
        SELECT l_linenumber, l_suppkey, l_orderkey, 
               l_partkey, l_quantity, l_extendedprice, l_discount, l_tax,
               l_returnflag, l_linestatus, l_shipdate, l_commitdate,
               l_receiptdate, l_shipinstruct, l_shipmode
        FROM lineitem
        ORDER BY l_linenumber, l_suppkey, l_orderkey
        "#,
```
Benchmark command
```sh
 cargo run --profile release-nonlto --bin dfbench -- sort-tpch -p /Users/yongting/Code/datafusion/benchmarks/data/tpch_sf10 -q 7 --memory-limit 1.2G
```
Notes:
- `target_partitions` config set to 14, and later configurations and results depend on this setting.
- For PR's benchmark runs, `sort_max_spill_merge_degree` is manually changed to 6, as a result: 
    - under 1.2G memory limit, 1 round of re-spill will be triggered
    - under 500M memory limit, 2 rounds of re-spill happens

#### Result
Main (1.2G): 
```
Q7 iteration 0 took 9374.7 ms and returned 59986052 rows
Q7 iteration 1 took 8117.6 ms and returned 59986052 rows
Q7 iteration 2 took 8549.1 ms and returned 59986052 rows
Q7 avg time: 8680.47 ms
```

Main (500M):
```
Fail with OOM
```

PR (1.2G):
```
ata/tpch_sf10 -q 7 --memory-limit 1G`
Q7 iteration 0 took 10723.6 ms and returned 59986052 rows
Q7 iteration 1 took 12962.8 ms and returned 59986052 rows
Q7 iteration 2 took 11739.7 ms and returned 59986052 rows
Q7 avg time: 11808.71 ms
```

PR (500M):
```
Q7 iteration 0 took 16233.1 ms and returned 59986052 rows
Q7 iteration 1 took 18568.4 ms and returned 59986052 rows
Q7 iteration 2 took 19173.4 ms and returned 59986052 rows
Q7 avg time: 17991.67 ms
```

File: datafusion/core/tests/memory_limit/mod.rs [hunk 8]
```diff
@@ -615,6 +616,104 @@ async fn test_disk_spill_limit_not_reached() -> Result<()> {
...
+// Test configuration `sort_max_spill_merge_degree` in external sorting
+// -------------------------------------------------------------------
+
+// Ensure invalid config value of `sort_max_spill_merge_degree` returns error
+#[rstest]
+#[case(0)]
+#[case(1)]
+#[tokio::test]
+async fn test_invalid_sort_max_spill_merge_degree(
+    #[case] sort_max_spill_merge_degree: usize,
```
Comment (qstommyshu) [inline datafusion/core/tests/memory_limit/mod.rs:628]: this #[case] syntax looks so elegant for writing repetitive tests

File: docs/source/user-guide/configs.md [hunk 9]
```diff
@@ -84,6 +84,7 @@ Environment variables are read during `SessionConfig` initialisation so they mus
...
+| datafusion.execution.sort_max_spill_merge_degree                        | 16                        | When doing external sorting, the maximum number of spilled files to read back at once. Those read files in the same merge step will be sort- preserving-merged and re-spilled, and the step will be repeated to reduce the number of spilled files in multiple passes, until a final sorted run can be produced.                                                                                                                                                                                                                                                         |
```
Comment (qstommyshu) [inline docs/source/user-guide/configs.md:87]: Great attention to detail, for updating the user guide!

**Comment by qstommyshu** [comment]
> ### Intended optimization
> If the memory pool is enough to hold more batches at a time (while `spill_max_spill_merge_degree` is still limited to 4, in case the merge-degree is too large and hurt performance in some cases) One additional config `sort_buffer_batch_capacity` is introduced, and set to `16` in the above example, the execution will look like: ...

Thanks for the clear explanation, that's a lot of great works, and looks really cool!

**Comment by alamb** [comment]
I plan to re-review this tomorrow

File: datafusion/common/src/config.rs [hunk 10]
```diff
@@ -337,6 +337,13 @@ config_namespace! {
         /// batches and merged.
         pub sort_in_place_threshold_bytes: usize, default = 1024 * 1024
 
+        /// When doing external sorting, the maximum number of spilled files to
+        /// read back at once. Those read files in the same merge step will be sort-
+        /// preserving-merged and re-spilled, and the step will be repeated to reduce
+        /// the number of spilled files in multiple passes, until a final sorted run
+        /// can be produced.
+        pub sort_max_spill_merge_degree: usize, default = 16
```
Comment (rluvaton) [inline datafusion/common/src/config.rs:345]: I have a concern about this that there can still be memory issue, if the batch from each stream together is above the memory limit

I have an implementation for this that is completely memory safe and will try to create a PR for that for inspiration

The way to decide on the degree is actually by storing for each spill file the largest amount of memory a single record batch taken, and then when deciding on the degree, you simply grow until you can no longer.

Comment (rluvaton) [reply datafusion/common/src/config.rs:345]: @2010YOUY01 and @alamb I hope before you merge this PR to look at https://github.com/apache/datafusion/pull/15700 to see what I mean

**Comment by rluvaton** [comment]
> > Also, to have a fully working larger than memory sort, you need to spill in
> > https://github.com/apache/datafusion/blob/362fcdfc7b9e00cb6126a0cbc41c9abb2637c563/datafusion/physical-plan/src/sorts/builder.rs#L74
> > 
> > In case the memory reservation is failing
> 
> Could you elaborate? I don't get it.

Look at the description for https://github.com/apache/datafusion/pull/15700 and the code for the PR it might clarify better

File: datafusion/common/src/config.rs (hunk 10 above)
Comment (2010YOUY01) [reply datafusion/common/src/config.rs:345]: > The reason why I'm picky about this is that it is a new configuration that will be hard to deprecate or change

This is a solid point, this option is intended to be manually set, and it has to ensure `(max_batch_size * per_partition_merge_degree * partition_count) < total_memory_limit`. If it's set correctly for a query, then the query should succeed.
The problem is the ever-growing number of configurations in DataFusion, and it seems impossible to set them all correctly. Enabling parallel merging optimization would require introducing yet another configuration, I'm also trying to avoid that (though too-many-configs problem might be a harsh reality we must accept).

**Comment by 2010YOUY01** [comment]
> > > Also, to have a fully working larger than memory sort, you need to spill in
> > > https://github.com/apache/datafusion/blob/362fcdfc7b9e00cb6126a0cbc41c9abb2637c563/datafusion/physical-plan/src/sorts/builder.rs#L74
> > > 
> > > In case the memory reservation is failing
> > 
> > 
> > Could you elaborate? I don't get it.
> 
> Maybe the description for #15700 might help

Thank you for providing an alternative approach.

I described my primary concern in https://github.com/apache/datafusion/pull/15700#discussion_r2041372025, I think it is not realistic to determine a batch’s memory size after a spilling roundtrip due to the implementation complexity. In such cases, if the estimation is off by a factor of 2, the actual memory usage could also increase by a factor of 2, which is not ideal.

**Comment by rluvaton** [comment]
Thank you, can you please take the fuzz test that I created in my pr and add it to yours, making sure it will pass (it will require you updating `row_hash.rs` file

**Comment by alamb** [comment]
> Benchmark results: (I think there is no significant regression for an extra round of re-spill, if it's running on a machine with fast SSDs)

It seems to me that there is a 30% regression in performance compared to main when there is enough memory, right?

> #### Result
> Main (1.2G):
> Q7 avg time: 8680.47 ms
> 
> PR (1.2G):
> Q7 avg time: 11808.71 ms


But this PR is significantly better that it can complete with only 500M of memory

Is there any way to regain the performance (maybe by choosing how many merge phases to do based on available memory rather than a fixed size)?

**Comment by alamb** [comment]
> Thank you, can you please take the fuzz test that I created in my pr and add it to yours, making sure it will pass (it will require you updating row_hash.rs file

@rluvaton  is there any way to make a PR with only the fuzz test in it (perhaps with some comments on what would pass/fail once we have this multi-pass algorithm

**Comment by rluvaton** [comment]
tested my fuzz tests with this pr and al of them are failing currently

File: datafusion/common/src/config.rs (hunk 10 above)
Comment (rluvaton) [reply datafusion/common/src/config.rs:345]: I'm keeping open my alternative approach as it seems working in limited memory envs as well (tested it locally with more data)

**Comment by 2010YOUY01** [comment]
> > Benchmark results: (I think there is no significant regression for an extra round of re-spill, if it's running on a machine with fast SSDs)
> 
> It seems to me that there is a 30% regression in performance compared to main when there is enough memory, right?
> 
> > #### Result
> > Main (1.2G):
> > Q7 avg time: 8680.47 ms
> > PR (1.2G):
> > Q7 avg time: 11808.71 ms
> 
> But this PR is significantly better that it can complete with only 500M of memory
> 
> Is there any way to regain the performance (maybe by choosing how many merge phases to do based on available memory rather than a fixed size)?

If we manually set this max merge degree to a larger value, the merging behavior will be equivalent to the current implementation:
```
Q7 iteration 0 took 7242.8 ms and returned 59986052 rows
Q7 iteration 1 took 7203.4 ms and returned 59986052 rows
Q7 iteration 2 took 9812.6 ms and returned 59986052 rows
Q7 avg time: 8086.24 ms
```

I think auto-tuning is possible, and is also a good future optimization to do, but it requires some work to extend the memory pool to estimate available memory for current reservation.

**Comment by 2010YOUY01** [comment]
> Thank you, can you please take the fuzz test that I created in my pr and add it to yours, making sure it will pass (it will require you updating `row_hash.rs` file

Those tests are great, but I think it's outside the scope of this PR: now external aggregation is using a different path for handling spills, the failures are not the regression caused by this PR.
It makes more sense to me to do it as a follow-up to: 1. Reuse spill handling code inside external aggr 2. make sure those tests pass.

**Comment by rluvaton** [comment]
In the PR that I created for fuzz tests that are also tests on sort which fails here as well
- #15727

**Comment by 2010YOUY01** [comment]
> tested my fuzz tests with this pr and all of them are failing currently

Update: I think the failure is not due to this PR's implementation, instead it's caused by `FairMemoryPool`'s limitation.

After manually setting `max_spill_merge_degree` to 2, first 3 tests passed, the 4th one failed with
```
Error: ResourcesExhausted("Additional allocation failed with top memory consumers (across reservations) as: mock_memory_consumer#2(can spill: false) consumed 1695480 bytes, ExternalSorterMerge[0]#1(can spill: false) consumed 401664 bytes. Error: Failed to allocate additional 297024 bytes for ExternalSorterMerge[0] with 0 bytes already allocated for this reservation - 8 bytes remain available for the total pool
```
I believe it's the limitation of `FairSpillPool` that non-spillable consumers are not able to back off, and it can block spilling consumers from normal execution. (and this specific test is also possible to pass due to some complex interactions if the runtime memory consumers are set up differently)

I'll try to come up with a minimal reproducer later.

**Comment by LogicFan** [comment]
I've tried to use this branch to sort data larger than memory. For 24GB parquet file, it produce error `Error: ArrowError(IoError("No space left on device (os error 28)", Os { code: 28, kind: StorageFull, message: "No space left on device" }), None)` (i think this suggest it uses up all 100GB default DiskManager?)

here is the configuration i use
```
    let cfg = SessionConfig::new()
        .with_sort_max_spill_merge_degree(2)
        .with_sort_spill_reservation_bytes(1 << 10)
        .with_sort_in_place_threshold_bytes(1 << 10)
        .set_usize("datafusion.execution.batch_size", 16)
        .set_usize("datafusion.execution.soft_max_rows_per_output_file", 2048)
        .set_usize("datafusion.execution.minimum_parallel_output_files", 32);

    let memory_pool = Arc::new(TrackConsumersPool::new(
        FairSpillPool::new(16 * (1 << 30)),
        NonZeroUsize::new(5).unwrap(),
    ));
```

**Comment by alamb** [comment]
> I've tried to use this branch to sort data larger than memory. For 24GB parquet file, it produce error `Error: ArrowError(IoError("No space left on device (os error 28)", Os { code: 28, kind: StorageFull, message: "No space left on device" }), None)` (i think this suggest it uses up all 100GB default DiskManager?)

We probably need to enable compression for such datasets (parquet is quite a bit more compressed than Arrow)

**Comment by ding-young** [comment]
This PR has a significant strength in that it works reliably even under a fairly conservative memory limit, which is impressive. I also learned a lot while reviewing it :). However, I do have the following concern about the approach based on user-configured `MAX_SPILL_MERGE_DEGREE`:

I don’t think adding a user-configurable merge degree is inherently a bad idea. The problem is that the appropriate value for `MAX_SPILL_MERGE_DEGREE` can vary significantly depending on the query being executed and the characteristics of the spilled `RecordBatches`. For example, merging many thin batches is very different from merging a few wide batches in terms of memory consumption, and the optimal degree will differ accordingly.

Unless the degree is defined in a way that consistently reflects some notion of aggressiveness or is based on actual memory consumption (e.g., in bytes), it would be hard to offer meaningful tuning guidance to users. In the end, as @rluvaton suggested, I think we do need to estimate something like `max_memory_bytes` per batch—even if the estimate isn’t very reliable.

Though relying on memory estimation may not be very reliable, it seems unavoidable in the context of multi-pass merge to estimate the memory size of the RecordBatches to be read from each spill file. This is because we need at least a rough estimate to decide how many spill files (or streams) can be read and merged simultaneously, so that we could automatically perform multi-pass merge without user's manual debugging.

What do you think about this? :thinking: 

Of course, I’m also currently trying to reproduce the case you pointed out in #15700, so if I discover a significant issue there, my opinion may very well change.

**Comment by rluvaton** [comment]
> Of course, I’m also currently trying to reproduce the case you pointed out in #15700, so if I discover a significant issue there, my opinion may very well change.

@ding-young  I already created a pr that reproduced the issues I outlined:
- #15727

**Comment by 2010YOUY01** [comment]
> This PR has a significant strength in that it works reliably even under a fairly conservative memory limit, which is impressive. I also learned a lot while reviewing it :). However, I do have the following concern about the approach based on user-configured `MAX_SPILL_MERGE_DEGREE`:
> 
> I don’t think adding a user-configurable merge degree is inherently a bad idea. The problem is that the appropriate value for `MAX_SPILL_MERGE_DEGREE` can vary significantly depending on the query being executed and the characteristics of the spilled `RecordBatches`. For example, merging many thin batches is very different from merging a few wide batches in terms of memory consumption, and the optimal degree will differ accordingly.
> 
> Unless the degree is defined in a way that consistently reflects some notion of aggressiveness or is based on actual memory consumption (e.g., in bytes), it would be hard to offer meaningful tuning guidance to users. In the end, as @rluvaton suggested, I think we do need to estimate something like `max_memory_bytes` per batch—even if the estimate isn’t very reliable.
> 
> Though relying on memory estimation may not be very reliable, it seems unavoidable in the context of multi-pass merge to estimate the memory size of the RecordBatches to be read from each spill file. This is because we need at least a rough estimate to decide how many spill files (or streams) can be read and merged simultaneously, so that we could automatically perform multi-pass merge without user's manual debugging.
> 
> What do you think about this? 🤔

I agree this knob would be very hard to be tuned manually. 

I was thinking the next step to be adding an `auto` option to the config, and automatically determine this max merge degree based on stats, according to avg batch size and memory budget. If it's wrong, the memory pool reports OOM, maybe the executor can retry with a small constant merge degree as the fallback path.

**Comment by 2010YOUY01** [comment]
> > tested my fuzz tests with this pr and all of them are failing currently
> 
> Update: I think the failure is not due to this PR's implementation, instead it's caused by `FairMemoryPool`'s limitation.
> 
> After manually setting `max_spill_merge_degree` to 2, first 3 tests passed, the 4th one failed with
> 
> ```
> Error: ResourcesExhausted("Additional allocation failed with top memory consumers (across reservations) as: mock_memory_consumer#2(can spill: false) consumed 1695480 bytes, ExternalSorterMerge[0]#1(can spill: false) consumed 401664 bytes. Error: Failed to allocate additional 297024 bytes for ExternalSorterMerge[0] with 0 bytes already allocated for this reservation - 8 bytes remain available for the total pool
> ```
> 
> I believe it's the limitation of `FairSpillPool` that non-spillable consumers are not able to back off, and it can block spilling consumers from normal execution. (and this specific test is also possible to pass due to some complex interactions if the runtime memory consumers are set up differently)
> 
> I'll try to come up with a minimal reproducer later.

I spotted a critical issue in this PR:

This approach works if the memory pool is only used by a single query, since we can guarantee during different stage in sort operator (partial sort -> SPM), the memory limit for a partition is roughly the same.

However, if a memory pool is shared among many queries (new query join the pool, get finished, and leave): given a certain partition, its memory budget can vary in its life cycle. For example: for a sort operator, when doing partial sorting it has 2GB memory budget, but when it moves to SPM phase, it only got 1GB because a new memory-consuming neighbor has recently joined.

As a result, SPM has to be inherently spillable.

I plan to close this PR and help to get https://github.com/apache/datafusion/pull/15700 merged, I have a rough idea to do a patch for my previous concern, I'll share my ideas there.

cc @rluvaton @ding-young
